File: frontend/src/api/downloadIntents.ts
// @ts-ignore
/* eslint-disable */
import { request, type RequestOptions } from "@/lib/request";

/** 按幂等键找回当前用户已提交的解析意图 GET /api/download-intents */
export async function findDownloadIntent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.findDownloadIntentParams,
  options?: RequestOptions
) {
  return request<API.ApiResponseIntentResponse_>("/api/download-intents", {
    method: "GET",
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** 提交持久解析意图 POST /api/download-intents */
export async function createDownloadIntent(
  body: API.IntentRequest,
  options?: RequestOptions
) {
  return request<API.ApiResponseIntentResponse_>("/api/download-intents", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** 查询当前用户的解析意图 GET /api/download-intents/${param0} */
export async function getDownloadIntent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getDownloadIntentParams,
  options?: RequestOptions
) {
  const { intent_id: param0, ...queryParams } = params;
  return request<API.ApiResponseIntentResponse_>(
    `/api/download-intents/${param0}`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 取消当前用户的解析意图 POST /api/download-intents/${param0}/cancel */
export async function cancelDownloadIntent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.cancelDownloadIntentParams,
  options?: RequestOptions
) {
  const { intent_id: param0, ...queryParams } = params;
  return request<API.ApiResponseIntentResponse_>(
    `/api/download-intents/${param0}/cancel`,
    {
      method: "POST",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 在原意图与剩余预算内更新过期解析结果 POST /api/download-intents/${param0}/refresh */
export async function refreshDownloadIntent(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.refreshDownloadIntentParams,
  options?: RequestOptions
) {
  const { intent_id: param0, ...queryParams } = params;
  return request<API.ApiResponseIntentResponse_>(
    `/api/download-intents/${param0}/refresh`,
    {
      method: "POST",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 分页查询当前用户的解析记录 GET /api/download-intents/history */
export async function listDownloadIntents(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listDownloadIntentsParams,
  options?: RequestOptions
) {
  return request<API.ApiResponseIntentHistoryResponse_>(
    "/api/download-intents/history",
    {
      method: "GET",
      params: {
        // limit has a default value: 20
        limit: "20",
        ...params,
      },
      ...(options || {}),
    }
  );
}

File: backend/app/api/routes/download_intents.py
from datetime import datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response

from app.api.admission import RateLimitAdmission
from app.api.deps import (
    IdempotencyKey,
    get_current_user,
    get_history_record_service,
    get_services,
    require_service,
)
from app.api.responses import ApiResponseRoute
from app.schemas.download_intents import (
    IntentHistoryResponse,
    IntentRequest,
    IntentResponse,
)
from app.schemas.history_records import HistoryRecordPageResponse
from app.services.auth.models import CurrentUser
from app.services.downloads.intents import IntentService
from app.services.history_records import (
    HistoryRecordCursor,
    HistoryRecordKind,
    HistoryRecordService,
)

router = APIRouter(
    route_class=ApiResponseRoute, prefix="/download-intents", tags=["download-intents"]
)


def get_intent_service(request: Request) -> IntentService:
    return require_service(get_services(request).intent_service, "parse intent")


User = Annotated[CurrentUser, Depends(get_current_user)]
Service = Annotated[IntentService, Depends(get_intent_service)]
HistoryService = Annotated[HistoryRecordService, Depends(get_history_record_service)]


@router.get(
    "",
    response_model=IntentResponse,
    operation_id="findDownloadIntent",
    summary="按幂等键找回当前用户已提交的解析意图",
)
async def find_intent(
    idempotency_key: Annotated[str, Query(min_length=1, max_length=128)],
    user: User,
    service: Service,
    response: Response,
) -> IntentResponse:
    response.headers["Cache-Control"] = "no-store"
    return IntentResponse.from_snapshot(
        await service.get_by_key(idempotency_key, user.owner_hash)
    )


@router.post(
    "",
    status_code=202,
    response_model=IntentResponse,
    operation_id="createDownloadIntent",
    dependencies=[Depends(RateLimitAdmission("inspect"))],
    summary="提交持久解析意图",
)
async def create_intent(
    body: IntentRequest,
    idempotency_key: IdempotencyKey,
    user: User,
    service: Service,
    response: Response,
) -> IntentResponse:
    result = await service.create(
        body.input, user.owner_hash, idempotency_key, quota=user.admission_quota
    )
    response.headers["Location"] = f"/api/download-intents/{result.id}"
    response.headers["Cache-Control"] = "no-store"
    return IntentResponse.from_snapshot(result)


@router.get(
    "/history",
    response_model=IntentHistoryResponse,
    operation_id="listDownloadIntents",
    summary="分页查询当前用户的解析记录",
)
async def list_intents(
    user: User,
    service: Service,
    response: Response,
    before: UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
) -> IntentHistoryResponse:
    response.headers["Cache-Control"] = "no-store"
    return IntentHistoryResponse.from_page(
        await service.history(user.owner_hash, before=before, limit=limit)
    )


@router.get(
    "/history/records",
    response_model=HistoryRecordPageResponse,
    operation_id="listHistoryRecords",
    summary="分页查询解析入口与视频内容分析记录",
)
async def list_history_records(
    user: User,
    service: HistoryService,
    response: Response,
    before_created_at: datetime | None = None,
    before_record_type: HistoryRecordKind | None = None,
    before_id: UUID | None = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
) -> HistoryRecordPageResponse:
    response.headers["Cache-Control"] = "no-store"
    cursor_parts = (before_created_at, before_record_type, before_id)
    if any(part is not None for part in cursor_parts) and not all(
        part is not None for part in cursor_parts
    ):
        raise HTTPException(
            status_code=422,
            detail=(
                "before_created_at, before_record_type and before_id "
                "are required together"
            ),
        )
    if before_created_at is None:
        before = None
    else:
        if before_record_type is None or before_id is None:
            raise HTTPException(
                status_code=422,
                detail="Incomplete history cursor",
            )
        before = HistoryRecordCursor(
            created_at=before_created_at,
            record_type=before_record_type,
            id=before_id,
        )
    return HistoryRecordPageResponse.from_page(
        await service.list(user.owner_hash, before=before, limit=limit)
    )


@router.get(
    "/{intent_id}",
    response_model=IntentResponse,
    operation_id="getDownloadIntent",
    summary="查询当前用户的解析意图",
)
async def get_intent(
    intent_id: UUID, user: User, service: Service, response: Response
) -> IntentResponse:
    response.headers["Cache-Control"] = "no-store"
    return IntentResponse.from_snapshot(await service.get(intent_id, user.owner_hash))


@router.post(
    "/{intent_id}/refresh",
    status_code=202,
    response_model=IntentResponse,
    operation_id="refreshDownloadIntent",
    dependencies=[Depends(RateLimitAdmission("inspect"))],
    summary="在原意图与剩余预算内更新过期解析结果",
)
async def refresh_intent(
    intent_id: UUID, user: User, service: Service, response: Response
) -> IntentResponse:
    response.headers["Cache-Control"] = "no-store"
    response.headers["Location"] = f"/api/download-intents/{intent_id}"
    return IntentResponse.from_snapshot(
        await service.refresh(intent_id, user.owner_hash, quota=user.admission_quota)
    )


@router.post(
    "/{intent_id}/cancel",
    response_model=IntentResponse,
    operation_id="cancelDownloadIntent",
    summary="取消当前用户的解析意图",
)
async def cancel_intent(
    intent_id: UUID, user: User, service: Service, response: Response
) -> IntentResponse:
    response.headers["Cache-Control"] = "no-store"
    return IntentResponse.from_snapshot(
        await service.cancel(intent_id, user.owner_hash)
    )
